Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp

```diff
@@ -238,21 +233,24 @@ void webkit_favicon_database_get_favicon(WebKitFaviconDatabase* database, const
  *
  * Finishes an operation started with webkit_favicon_database_get_favicon().
  *
- * Returns: (transfer full): a new reference to a #cairo_surface_t, or
- * %NULL in case of error.
+ * Returns: (transfer full): a new favicon image, or %NULL in case of error.
  */
+#if USE(GTK4)
+GdkTexture* webkit_favicon_database_get_favicon_finish(WebKitFaviconDatabase* database, GAsyncResult* result, GError** error)
+#else
 cairo_surface_t* webkit_favicon_database_get_favicon_finish(WebKitFaviconDatabase* database, GAsyncResult* result, GError** error)
+#endif
 {
-    g_return_val_if_fail(WEBKIT_IS_FAVICON_DATABASE(database), 0);
-    g_return_val_if_fail(g_task_is_valid(result, database), 0);
-
-    GTask* task = G_TASK(result);
-    if (auto* icon = g_task_propagate_pointer(task, error))
-        return static_cast<cairo_surface_t*>(icon);
+    g_return_val_if_fail(WEBKIT_IS_FAVICON_DATABASE(database), nullptr);
+    g_return_val_if_fail(g_task_is_valid(result, database), nullptr);
 
-    return nullptr;
-}
+#if USE(GTK4)
+    auto icon = adoptRef(static_cast<cairo_surface_t*>(g_task_propagate_pointer(G_TASK(result), error)));
+    return icon ? cairoSurfaceToGdkTexture(icon.get()) : nullptr;
```

Comment:
```suggestion
    return icon ? cairoSurfaceToGdkTexture(icon.get()).leakRef() : nullptr;
```

Yes really... even though the smart pointer kind of gets in the way here, it's important to use them to transfer ownership wherever possible, especially when the function is defined in an entirely different file.